File: skills/inspect-and-debug/harness-doctor/scripts/lib/scan.ts
/**
 * Walk the transcripts root and turn every transcript in the window into the
 * five fact tables.
 *
 * Two things the tree teaches you: transcripts nest, and most of them are
 * nested. A one-level glob finds only the main sessions, which was ~10% of the
 * reference corpus (386 of 3,101): the rest sit under
 * `<sid>/subagents/.../agent-<id>.jsonl`. So the walk is recursive, and it skips
 * `journal.jsonl` and `*.meta.json`.
 *
 * Memory: one file is read, converted to rows, and released before the next one
 * opens. Nothing holds the raw corpus.
 */
import { readdirSync, readFileSync, statSync, mkdirSync, writeFileSync, appendFileSync } from "node:fs";
import type { Dirent } from "node:fs";
import { join } from "node:path";
import type {
  AuditOptions,
  BashRow,
  ErrorRow,
  FactTables,
  ScanResult,
  SessionKind,
  SessionRow,
  ToolRow,
  UserMsgRow,
} from "./types.ts";
import { normalizeCommand, errorSignature } from "./normalize.ts";
import { redactIf } from "./redact.ts";

/** A transcript file located on disk, before it is parsed. */
interface TranscriptRef {
  file: string;
  proj: string;
  sid: string;
  kind: SessionKind;
  parent?: string;
  run?: string;
}

/** Rows produced by one transcript. */
interface FileRows {
  session: SessionRow;
  tools: ToolRow[];
  bash: BashRow[];
  userMsgs: UserMsgRow[];
  errors: ErrorRow[];
}

const MAX_TEXT = 4000;
const MAX_ERR = 600;
const MAX_OUT_OK = 200;
const MAX_OUT_ERR = 600;
const MAX_ARG = 500;

/** Harness-generated wrappers that are not human input. */
const HARNESS_BLOCKS = [
  /<system-reminder>[\s\S]*?<\/system-reminder>/g,
  /<task-notification>[\s\S]*?<\/task-notification>/g,
  /<command-name>[\s\S]*?<\/command-name>/g,
  /<command-message>[\s\S]*?<\/command-message>/g,
  /<command-args>[\s\S]*?<\/command-args>/g,
  /<local-command-stdout>[\s\S]*?<\/local-command-stdout>/g,
  /<local-command-stderr>[\s\S]*?<\/local-command-stderr>/g,
];

/** Openers that mark a turn as machine-generated even when unclosed. */
const HARNESS_PREFIX =
  /^(?:\[Request interrupted|<system-reminder|<task-notification|<command-name|<command-message|<command-args|<local-command|Caveat: The messages below|API Error|\[Tool result)/;

/** ISO timestamp to epoch seconds, or undefined when absent or unparseable. */
const epoch = (ts: unknown) => {
  if (typeof ts !== "string") return undefined;
  const ms = Date.parse(ts);
  return Number.isFinite(ms) ? Math.floor(ms / 1000) : undefined;
};

/** ISO timestamp to epoch seconds with millisecond precision, for durations. */
const epochMs = (ts: unknown) => {
  if (typeof ts !== "string") return undefined;
  const ms = Date.parse(ts);
  return Number.isFinite(ms) ? ms : undefined;
};

/** Anything to a string: strings pass through, objects are JSON. */
const str = (v: unknown) => (typeof v === "string" ? v : v == null ? "" : JSON.stringify(v));

/** Flatten a tool_result `content` field: a string, or text blocks in an array. */
const resultText = (content: unknown): string => {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((b) => {
        const block = b as Record<string, unknown>;
        if (block?.type === "text") return str(block.text);
        if (block?.type === "image") return "";
        return "";
      })
      .join("\n");
  }
  return content == null ? "" : str(content);
};

/** The tool-specific argument that identifies what a call touched. */
const primaryArg = (tool: string, input: Record<string, unknown>) => {
  const pick = (...keys: string[]) => {
    for (const k of keys) if (typeof input[k] === "string" && input[k]) return input[k] as string;
    return undefined;
  };
  switch (tool) {
    case "Read":
    case "Edit":
    case "Write":
    case "MultiEdit":
    case "NotebookEdit":
      return pick("file_path", "notebook_path", "path");
    case "Grep":
    case "Glob":
      return pick("pattern");
    case "WebFetch":
      return pick("url");
    case "WebSearch":
      return pick("query");
    case "Agent":
    case "Task":
      return pick("subagent_type", "description");
    case "Skill":
      return pick("skill", "command");
    default:
      return undefined;
  }
};

/** Cut a string to `n` chars. */
const clip = (s: string, n: number) => (s.length > n ? s.slice(0, n) : s);

/**
 * Classify a transcript by where it sits under the project dir: depth 1 is a
 * main session, a `subagents/workflows/wf_<id>` ancestor makes it a workflow
 * agent, anything else nested is a plain subagent.
 */
const classify = (parts: string[]): Pick<TranscriptRef, "kind" | "parent" | "run"> => {
  if (parts.length <= 1) return { kind: "main" };
  const parent = parts[0];
  const wfIndex = parts.findIndex((p) => p.startsWith("wf_"));
  if (wfIndex > 0 && parts[wfIndex - 1] === "workflows") {
    return { kind: "workflow-agent", parent, run: parts[wfIndex] };
  }
  return { kind: "subagent", parent };
};

/**
 * List every transcript under the root, recursively, sorted by path so two runs
 * see the same order. `projects` filters project dirs by substring.
 */
export const listTranscripts = (root: string, projects: string[] = []) => {
  const found: TranscriptRef[] = [];
  let projectDirs: string[] = [];
  try {
    projectDirs = readdirSync(root, { withFileTypes: true })
      .filter((e) => e.isDirectory())
      .map((e) => e.name)
      .sort();
  } catch {
    return found;
  }
  for (const proj of projectDirs) {
    if (projects.length > 0 && !projects.some((p) => proj.includes(p))) continue;
    const projRoot = join(root, proj);
    const walk = (dir: string, parts: string[]) => {
      let entries: Dirent[];
      try {
        entries = readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }
      for (const entry of [...entries].sort((a, b) => a.name.localeCompare(b.name))) {
        if (entry.isDirectory()) {
          walk(join(dir, entry.name), [...parts, entry.name]);
          continue;
        }
        if (!entry.isFile() || !entry.name.endsWith(".jsonl")) continue;
        if (entry.name === "journal.jsonl") continue;
        found.push({
          file: join(dir, entry.name),
          proj,
          sid: entry.name.slice(0, -".jsonl".length),
          ...classify([...parts, entry.name]),
        });
      }
    };
    walk(projRoot, []);
  }
  return found.sort((a, b) => a.file.localeCompare(b.file));
};

/** Strip harness-generated blocks so a correction scan sees only human words. */
const stripHarness = (text: string) => {
  let out = text;
  for (const re of HARNESS_BLOCKS) out = out.replace(re, " ");
  return out.trim();
};

/** One in-flight tool call, waiting for its result. */
interface PendingCall {
  tool: string;
  input: Record<string, unknown>;
  tMs?: number;
  side: boolean;
}

/**
 * Parse one transcript file into fact rows. Malformed lines are skipped
 * silently: a live session's last line is often half-written.
 */
export const parseTranscript = (ref: TranscriptRef, bytes: number, options: AuditOptions): FileRows => {
  const { redact } = options;
  const tools: ToolRow[] = [];
  const bash: BashRow[] = [];
  const userMsgs: UserMsgRow[] = [];
  const errors: ErrorRow[] = [];

  const pending = new Map<string, PendingCall>();
  const seenResults = new Set<string>();
  const seenTurns = new Set<string>();
  const models: Record<string, number> = {};
  const toolCounts: Record<string, number> = {};

  const session: SessionRow = {
    sid: ref.sid,
    proj: ref.proj,
    file: ref.file,
    bytes,
    kind: ref.kind,
    parent: ref.parent,
    run: ref.run,
    assistantMsgs: 0,
    userMsgs: 0,
    inTokens: 0,
    outTokens: 0,
    cacheReadTokens: 0,
    cacheCreateTokens: 0,
    maxCtx: 0,
    firstTurnCacheCreate: 0,
    models,
    tools: toolCounts,
    interrupts: 0,
    errors: 0,
  };
  let firstTurnSeen = false;

  let raw = "";
  try {
    raw = readFileSync(ref.file, "utf8");
  } catch {
    return { session, tools, bash, userMsgs, errors };
  }

  /** Record one user turn, flagging whether a human really typed it. */
  const pushUserMsg = (textRaw: string, ctx: { t?: number; side: boolean; isMeta: boolean }) => {
    if (textRaw.includes("[Request interrupted")) session.interrupts++;
    const stripped = stripHarness(textRaw);
    const body = stripped || textRaw.trim();
    const human =
      ref.kind === "main" && !ctx.side && !ctx.isMeta && stripped.length > 0 && !HARNESS_PREFIX.test(stripped);
    userMsgs.push({
      sid: ref.sid,
      proj: ref.proj,
      kind: ref.kind,
      t: ctx.t,
      text: redactIf(redact, clip(body, MAX_TEXT)),
      side: ctx.side,
      human,
    });
  };

  /** Record one tool call: a ToolRow always, plus a BashRow and an ErrorRow when due. */
  const emitCall = (c: {
    tool: string;
    input: Record<string, unknown>;
    t?: number;
    durSec?: number;
    err: boolean;
    text: string;
    side: boolean;
  }) => {
    const outChars = c.text.length;
    const arg = primaryArg(c.tool, c.input);
    tools.push({
      sid: ref.sid,
      proj: ref.proj,
      kind: ref.kind,
      t: c.t,
      tool: c.tool,
      durSec: c.durSec,
      err: c.err,
      outChars,
      arg: arg ? clip(arg, MAX_ARG) : undefined,
      side: c.side,
    });
    if (c.tool === "Bash") {
      const cmd = str(c.input.command ?? c.input.cmd ?? "");
      const n = normalizeCommand(cmd);
      bash.push({
        sid: ref.sid,
        proj: ref.proj,
        kind: ref.kind,
        t: c.t,
        durSec: c.durSec,
        err: c.err,
        outChars,
        cmd: redactIf(redact, cmd),
        desc: typeof c.input.description === "string" ? c.input.description : undefined,
        family: n.family,
        bin: n.bin,
        bg: c.input.run_in_background === true || n.background,
        out: redactIf(redact, clip(c.text, c.err ? MAX_OUT_ERR : MAX_OUT_OK)),
        side: c.side,
      });
    }
    if (!c.err) return;
    const msgText = redactIf(redact, clip(c.text, MAX_ERR));
    errors.push({
      sid: ref.sid,
      proj: ref.proj,
      kind: ref.kind,
      t: c.t,
      tool: c.tool,
      input: redactIf(redact, clip(JSON.stringify(c.input), MAX_ERR)),
      msg: msgText,
      sig: errorSignature(msgText),
    });
  };

  for (const line of raw.split("\n")) {
    if (!line.trim()) continue;
    let o: Record<string, unknown>;
    try {
      o = JSON.parse(line) as Record<string, unknown>;
    } catch {
      continue;
    }
    const type = String(o.type ?? "");
    const t = epoch(o.timestamp);
    const tMs = epochMs(o.timestamp);
    if (t !== undefined) {
      if (session.start === undefined) session.start = t;
      session.end = t;
    }
    if (typeof o.cwd === "string" && !session.cwd) session.cwd = o.cwd;
    if (typeof o.gitBranch === "string" && o.gitBranch) session.branch = o.gitBranch;
    if (typeof o.version === "string") session.ver = o.version;
    const side = o.isSidechain === true;
    const msg = (o.message ?? {}) as Record<string, unknown>;

    if (type === "assistant") {
      session.assistantMsgs++;
      const model = String(msg.model ?? "unknown");
      const usage = (msg.usage ?? {}) as Record<string, number>;
      const turnKey = String(o.requestId ?? msg.id ?? `line-${session.assistantMsgs}`);
      if (!seenTurns.has(turnKey) && msg.usage) {
        seenTurns.add(turnKey);
        const inTok = usage.input_tokens ?? 0;
        const outTok = usage.output_tokens ?? 0;
        const cr = usage.cache_read_input_tokens ?? 0;
        const cc = usage.cache_creation_input_tokens ?? 0;
        session.inTokens += inTok;
        session.outTokens += outTok;
        session.cacheReadTokens += cr;
        session.cacheCreateTokens += cc;
        session.maxCtx = Math.max(session.maxCtx, inTok + cr + cc);
        models[model] = (models[model] ?? 0) + 1;
        if (!firstTurnSeen) {
          session.firstTurnCacheCreate = cc;
          firstTurnSeen = true;
        }
      }
      for (const b of (msg.content ?? []) as Array<Record<string, unknown>>) {
        if (b?.type !== "tool_use") continue;
        const tool = String(b.name ?? "tool");
        const id = typeof b.id === "string" ? b.id : "";
        const input = (b.input ?? {}) as Record<string, unknown>;
        toolCounts[tool] = (toolCounts[tool] ?? 0) + 1;
        if (id) pending.set(id, { tool, input, tMs, side });
      }
      continue;
    }

    if (type !== "user") continue;
    session.userMsgs++;
    const tur = (o.toolUseResult ?? {}) as Record<string, unknown>;
    if (tur?.interrupted === true) session.interrupts++;
    const content = msg.content;

    if (typeof content === "string") {
      pushUserMsg(content, { t, side, isMeta: o.isMeta === true });
      continue;
    }
    if (!Array.isArray(content)) continue;

    for (const b of content as Array<Record<string, unknown>>) {
      if (b?.type === "text") {
        pushUserMsg(str(b.text), { t, side, isMeta: o.isMeta === true });
        continue;
      }
      if (b?.type !== "tool_result") continue;
      const id = typeof b.tool_use_id === "string" ? b.tool_use_id : "";
      if (id && seenResults.has(id)) continue;
      if (id) seenResults.add(id);
      const call = id ? pending.get(id) : undefined;
      if (id) pending.delete(id);
      let text = resultText(b.content);
      if (!text && (typeof tur.stdout === "string" || typeof tur.stderr === "string")) {
        text = `${str(tur.stdout)}${str(tur.stderr)}`;
      }
      if (text.includes("[Request interrupted")) session.interrupts++;
      const err = b.is_error === true;
      if (err) session.errors++;
      const tool = call?.tool ?? "unknown";
      const input = call?.input ?? {};
      const callT = call?.tMs !== undefined ? Math.floor(call.tMs / 1000) : t;
      const durSec =
        call?.tMs !== undefined && tMs !== undefined
          ? Math.max(0, Math.round(((tMs - call.tMs) / 1000) * 1000) / 1000)
          : undefined;
      emitCall({ tool, input, t: callT, durSec, err, text, side: call?.side ?? side });
    }
  }

  // Tool calls whose result never arrived: the session ended or was interrupted.
  for (const [, call] of pending) {
    emitCall({
      tool: call.tool,
      input: call.input,
      t: call.tMs !== undefined ? Math.floor(call.tMs / 1000) : undefined,
      durSec: undefined,
      err: false,
      text: "",
      side: call.side,
    });
  }

  return { session, tools, bash, userMsgs, errors };
};

/**
 * Scan every transcript in the window and return the fact tables plus counts.
 * Progress goes to stderr so stdout stays a clean summary.
 */
export const scanTranscripts = (options: AuditOptions): ScanResult => {
  const toEpoch = Math.floor(Date.now() / 1000);
  const fromEpoch = toEpoch - options.days * 86400;
  const refs = listTranscripts(options.root, options.projects);
  const tables: FactTables = { sessions: [], tools: [], bash: [], userMsgs: [], errors: [] };
  const seenProjects = new Set<string>();
  let scanned = 0;
  let skipped = 0;

  for (let i = 0; i < refs.length; i++) {
    const ref = refs[i]!;
    let bytes = 0;
    let mtime = 0;
    try {
      const st = statSync(ref.file);
      bytes = st.size;
      mtime = Math.floor(st.mtimeMs / 1000);
    } catch {
      skipped++;
      continue;
    }
    // A file untouched since before the window cannot hold activity inside it.
    if (mtime < fromEpoch) {
      skipped++;
      continue;
    }
    const rows = parseTranscript(ref, bytes, options);
    const end = rows.session.end ?? mtime;
    if (end < fromEpoch) {
      skipped++;
      continue;
    }
    scanned++;
    seenProjects.add(ref.proj);
    tables.sessions.push(rows.session);
    tables.tools.push(...rows.tools);
    tables.bash.push(...rows.bash);
    tables.userMsgs.push(...rows.userMsgs);
    tables.errors.push(...rows.errors);
    if (scanned % 200 === 0 || i === refs.length - 1) {
      process.stderr.write(
        `scanned ${scanned}/${refs.length} files, ${tables.tools.length} tool calls, ${tables.bash.length} bash\n`,
      );
    }
  }
  process.stderr.write(`scan done: ${scanned} transcripts, ${skipped} skipped, ${seenProjects.size} projects\n`);

  return {
    tables,
    filesScanned: scanned,
    filesSkipped: skipped,
    projectCount: seenProjects.size,
    fromEpoch,
    toEpoch,
  };
};

/** File name each table lands in, in the order they are written. */
const TABLE_FILES: Array<[keyof FactTables, string]> = [
  ["sessions", "sessions.jsonl"],
  ["tools", "tools.jsonl"],
  ["bash", "bash.jsonl"],
  ["userMsgs", "user-msgs.jsonl"],
  ["errors", "errors.jsonl"],
];

/**
 * Write the five JSONL fact tables into `outDir`, creating it if needed.
 * Returns the paths written. Rows go out in scan order, so two runs over the
 * same transcripts produce byte-identical files.
 */
export const writeTables = (tables: FactTables, outDir: string) => {
  mkdirSync(outDir, { recursive: true });
  const written: string[] = [];
  for (const [key, name] of TABLE_FILES) {
    const path = join(outDir, name);
    writeFileSync(path, "", "utf8");
    const rows: unknown[] = tables[key];
    let chunk = "";
    for (const row of rows) {
      chunk += `${JSON.stringify(row)}\n`;
      if (chunk.length > 4_000_000) {
        appendFileSync(path, chunk, "utf8");
        chunk = "";
      }
    }
    if (chunk) appendFileSync(path, chunk, "utf8");
    written.push(path);
  }
  return written;
};

File: skills/inspect-and-debug/harness-doctor/scripts/lib/scan.py
"""Transcript discovery and per-file parsing.

Walks the transcripts root recursively (a `*/*.jsonl` glob only finds main
sessions; subagents and workflow agents live nested under
`<parent-sid>/subagents/**/`), then parses each file one line at a time so
only one transcript's raw JSON is ever in memory at once.
"""

import json
import os
import sys
import time
from collections import Counter
from datetime import datetime, timezone
from pathlib import Path

from .normalize import is_human_text, normalize_command, normalize_sig
from .redact import redact_text

_TOOL_RESULT_CONTENT_CAP = 2000
_INTERRUPT_MARK = "[Request interrupted"


def parse_ts(d):
    """Epoch seconds from a transcript line's timestamp field, or None."""
    raw = d.get("timestamp")
    if not raw:
        return None
    try:
        return datetime.fromisoformat(str(raw).replace("Z", "+00:00")).timestamp()
    except (ValueError, TypeError):
        return None


def _classify(dir_parts):
    """kind, parent, run from the directory chain relative to the root."""
    if len(dir_parts) <= 1:
        return "main", None, None
    parent = dir_parts[1]
    kind = "subagent"
    run = None
    for comp in dir_parts[2:]:
        if comp.startswith("wf_"):
            kind = "workflow-agent"
            run = comp
            break
    return kind, parent, run


def _primary_arg(name, inp):
    """Tool-specific primary argument used for ToolRow.arg."""
    if name in ("Read", "Edit", "Write", "NotebookEdit"):
        return inp.get("file_path") or inp.get("path")
    if name in ("Grep", "Glob"):
        return inp.get("pattern")
    if name == "WebFetch":
        return inp.get("url")
    if name == "WebSearch":
        return inp.get("query")
    if name in ("Task", "Agent"):
        return inp.get("subagent_type")
    return None


def _extract_user_content(cont, pend, t, sid, proj, kind, side, tool_rows, bash_rows, err_rows, redact_on):
    """Handle one user-turn's content: join text, join tool_results to their
    pending tool_use, and emit ToolRow/BashRow/ErrorRow rows as a side effect.
    Returns (text, interrupted).
    """
    if isinstance(cont, str):
        return cont, _INTERRUPT_MARK in cont
    if not isinstance(cont, list):
        return "", False

    interrupted = False
    textparts = []
    for b in cont:
        if not isinstance(b, dict):
            continue
        btype = b.get("type")
        if btype == "text":
            textparts.append(b.get("text") or "")
            continue
        if btype != "tool_result":
            continue
        name, inp, t0 = pend.pop(b.get("tool_use_id"), (None, {}, None))
        content = b.get("content")
        if isinstance(content, list):
            content = " ".join(x.get("text", "") for x in content if isinstance(x, dict))
        content = content if isinstance(content, str) else json.dumps(content)[:_TOOL_RESULT_CONTENT_CAP]
        if _INTERRUPT_MARK in content:
            interrupted = True
        is_err = bool(b.get("is_error"))
        dur = (t - t0) if (t is not None and t0 is not None) else None

        if name:
            tool_row = {"sid": sid, "proj": proj, "kind": kind, "tool": name,
                        "err": is_err, "outChars": len(content or ""), "side": side}
            if t0 is not None:
                tool_row["t"] = t0
            if dur is not None:
                tool_row["durSec"] = round(dur, 3)
            arg = _primary_arg(name, inp)
            if arg:
                tool_row["arg"] = arg
            tool_rows.append(tool_row)

            if name == "Bash":
                family, binname = normalize_command(inp.get("command") or "")
                out_cap = 600 if is_err else 200
                bash_row = {"sid": sid, "proj": proj, "kind": kind,
                            "cmd": redact_text((inp.get("command") or "")[:1500], redact_on),
                            "family": family, "bin": binname,
                            "bg": bool(inp.get("run_in_background")),
                            "err": is_err, "outChars": len(content or ""), "side": side,
                            "out": redact_text((content or "")[:out_cap], redact_on)}
                if inp.get("description"):
                    bash_row["desc"] = str(inp.get("description"))[:200]
                if t0 is not None:
                    bash_row["t"] = t0
                if dur is not None:
                    bash_row["durSec"] = round(dur, 3)
                bash_rows.append(bash_row)

        if is_err:
            err_row = {"sid": sid, "proj": proj, "kind": kind,
                       "input": redact_text(json.dumps(inp)[:600], redact_on),
                       "msg": redact_text((content or "")[:600], redact_on),
                       "sig": normalize_sig(content or "")}
            if name:
                err_row["tool"] = name
            if t0 is not None:
                err_row["t"] = t0
            err_rows.append(err_row)

    text = "\n".join(x for x in textparts if x).strip()
    if _INTERRUPT_MARK in text:
        interrupted = True
    return text, interrupted


def parse_transcript(path, sid, proj, kind, parent, run, file_bytes, redact_on):
    """Parse one transcript file into (SessionRow or None, tool rows, bash
    rows, user-msg rows, error rows, per-model token usage).

    The per-model usage dict is exact (summed per assistant message) and is
    used only for the tokens.byModel aggregate; it is not part of the
    SessionRow fact-table schema.
    """
    start = end = None
    cwd = branch = ver = None
    assistant_msgs = user_msgs_count = interrupts = 0
    in_tok = out_tok = cr_tok = cc_tok = max_ctx = 0
    first_turn_cache_create = None
    models = Counter()
    tools = Counter()
    model_usage = {}
    pend = {}
    tool_rows, bash_rows, user_rows, err_rows = [], [], [], []

    try:
        fh = open(path, "r", encoding="utf-8", errors="replace")
    except OSError:
        return None, [], [], [], [], {}

    with fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            try:
                d = json.loads(line)
            except (json.JSONDecodeError, ValueError):
                continue
            if not isinstance(d, dict):
                continue

            t = parse_ts(d)
            if t is not None:
                start = t if start is None else min(start, t)
                end = t if end is None else max(end, t)
            cwd = d.get("cwd") or cwd
            branch = d.get("gitBranch") or branch
            ver = d.get("version") or ver
            side = bool(d.get("isSidechain"))
            is_meta = bool(d.get("isMeta"))
            typ = d.get("type")
            m = d.get("message") or {}
            cont = m.get("content")

            if typ == "assistant":
                assistant_msgs += 1
                model = m.get("model")
                usage = m.get("usage") or {}
                it = usage.get("input_tokens", 0) or 0
                ot = usage.get("output_tokens", 0) or 0
                crt = usage.get("cache_read_input_tokens", 0) or 0
                cct = usage.get("cache_creation_input_tokens", 0) or 0
                if model:
                    models[model] += 1
                    mu = model_usage.setdefault(model, {"in": 0, "out": 0, "cacheRead": 0, "cacheCreate": 0})
                    mu["in"] += it
                    mu["out"] += ot
                    mu["cacheRead"] += crt
                    mu["cacheCreate"] += cct
                in_tok += it
                out_tok += ot
                cr_tok += crt
                cc_tok += cct
                max_ctx = max(max_ctx, it + crt + cct)
                if first_turn_cache_create is None:
                    first_turn_cache_create = cct
                if isinstance(cont, list):
                    for b in cont:
                        if isinstance(b, dict) and b.get("type") == "tool_use":
                            name = b.get("name")
                            tools[name] += 1
                            pend[b.get("id")] = (name, b.get("input") or {}, t)

            elif typ == "user":
                text, interrupted = _extract_user_content(
                    cont, pend, t, sid, proj, kind, side,
                    tool_rows, bash_rows, err_rows, redact_on,
                )
                tur = d.get("toolUseResult")
                if isinstance(tur, dict) and tur.get("interrupted"):
                    interrupted = True
                if interrupted:
                    interrupts += 1
                if text:
                    user_msgs_count += 1
                    row = {"sid": sid, "proj": proj, "kind": kind,
                           "text": redact_text(text[:4000], redact_on),
                           "side": side,
                           "human": is_human_text(text, kind, side, is_meta)}
                    if t is not None:
                        row["t"] = t
                    user_rows.append(row)

    if assistant_msgs == 0 and user_msgs_count == 0:
        return None, tool_rows, bash_rows, user_rows, err_rows, model_usage

    session = {"sid": sid, "proj": proj, "file": str(path), "bytes": file_bytes, "kind": kind}
    if cwd:
        session["cwd"] = cwd
    if branch:
        session["branch"] = branch
    if ver:
        session["ver"] = ver
    if start is not None:
        session["start"] = start
    if end is not None:
        session["end"] = end
    if parent:
        session["parent"] = parent
    if run:
        session["run"] = run
    session.update({
        "assistantMsgs": assistant_msgs,
        "userMsgs": user_msgs_count,
        "inTokens": in_tok,
        "outTokens": out_tok,
        "cacheReadTokens": cr_tok,
        "cacheCreateTokens": cc_tok,
        "maxCtx": max_ctx,
        "firstTurnCacheCreate": first_turn_cache_create or 0,
        "models": dict(models),
        "tools": dict(tools),
        "interrupts": interrupts,
        "errors": sum(1 for r in tool_rows if r.get("err")),
    })
    return session, tool_rows, bash_rows, user_rows, err_rows, model_usage


def _iter_candidates(root: Path):
    """All *.jsonl transcript candidates under root, deterministically
    ordered, skipping journal.jsonl (*.meta.json is a different extension and
    is never matched by the *.jsonl walk).
    """
    found = []
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames.sort()
        for fn in filenames:
            if fn.endswith(".jsonl") and fn != "journal.jsonl":
                found.append(Path(dirpath) / fn)
    found.sort(key=lambda p: str(p))
    return found


def scan_all(root, days, project_filters, redact_on, progress_stream=None):
    """Walk root and parse every in-window transcript.

    Returns (sessions, tools, bashRows, userMsgs, errors, stats) where stats
    has filesScanned, filesSkipped, fromTs, toTs, modelTokens (exact per-model
    token totals, used only for the tokens.byModel aggregate).
    """
    progress_stream = sys.stderr if progress_stream is None else progress_stream
    root = Path(root)
    now = time.time()
    cutoff = now - days * 86400
    project_filters = project_filters or []

    candidates = _iter_candidates(root) if root.exists() else []
    total = len(candidates)

    sessions, tools, bash_rows, user_msgs, errors = [], [], [], [], []
    model_tokens = {}
    files_scanned = files_skipped = 0

    for idx, path in enumerate(candidates, 1):
        try:
            rel_parts = path.relative_to(root).parts
        except ValueError:
            files_skipped += 1
            continue
        if len(rel_parts) < 2:
            files_skipped += 1
            continue
        proj = rel_parts[0]
        if project_filters and not any(pf in proj for pf in project_filters):
            files_skipped += 1
            continue
        try:
            st = path.stat()
        except OSError:
            files_skipped += 1
            continue
        if st.st_mtime < cutoff:
            files_skipped += 1
            continue

        kind, parent, run = _classify(rel_parts[:-1])
        sid = path.name[:-6] if path.name.endswith(".jsonl") else path.name

        session, t_rows, b_rows, u_rows, e_rows, mu = parse_transcript(
            path, sid, proj, kind, parent, run, st.st_size, redact_on,
        )
        files_scanned += 1
        if session is not None:
            sessions.append(session)
        tools.extend(t_rows)
        bash_rows.extend(b_rows)
        user_msgs.extend(u_rows)
        errors.extend(e_rows)
        for model, counts in mu.items():
            acc = model_tokens.setdefault(model, {"in": 0, "out": 0, "cacheRead": 0, "cacheCreate": 0})
            for k, v in counts.items():
                acc[k] += v

        if idx % 100 == 0 or idx == total:
            print(f"scanned {idx}/{total} files...", file=progress_stream)

    stats = {"filesScanned": files_scanned, "filesSkipped": files_skipped, "fromTs": cutoff, "toTs": now,
              "modelTokens": model_tokens}
    return sessions, tools, bash_rows, user_msgs, errors, stats
